Sources/AsyncHTTPClient/RequestBag.swift

```diff
@@ -0,0 +1,843 @@
+            case .finished(error: .some(let error)):
+                return .failFuture(error)
+            case .finished(error: .none):
+                preconditionFailure("A write was made, after the request has completed")
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        switch action {
+        case .failTask(let error):
+            if self.task.eventLoop.inEventLoop {
+                self.delegate.didReceiveError(task: self.task, error)
+                self.task.fail(with: error, delegateType: Delegate.self)
+            } else {
+                self.task.eventLoop.execute {
+                    self.delegate.didReceiveError(task: self.task, error)
+                    self.task.fail(with: error, delegateType: Delegate.self)
+                }
+            }
+            return self.task.eventLoop.makeFailedFuture(error)
+
+        case .failFuture(let error):
+            return self.task.eventLoop.makeFailedFuture(error)
+
+        case .write(let part, let writer, let future):
+            writer.writeRequestBodyPart(part, request: self)
+            self.didSendRequestPart(part)
+            return future
+        }
+    }
+
+    enum FinishAction {
+        case forwardStreamFinished(HTTPRequestExecutor, [EventLoopPromise<Void>]?)
+        case forwardStreamFailureAndFailTask(HTTPRequestExecutor, Error, [EventLoopPromise<Void>]?)
+        case none
+    }
+
+    func finishRequestBodyStream(_ result: Result<Void, Error>) {
+        let action = self.stateLock.withLock { () -> FinishAction in
+            switch self._state {
+            case .initialized, .queued:
+                preconditionFailure("Invalid state: \(self._state)")
+            case .executing(let executor, let requestState, let responseState):
+                switch requestState {
+                case .initialized:
+                    preconditionFailure("Request stream must be started before it can be finished")
+                case .producing:
+                    switch result {
+                    case .success:
+                        self._state = .executing(executor, .finished, responseState)
+                        return .forwardStreamFinished(executor, nil)
+                    case .failure(let error):
+                        self._state = .finished(error: error)
+                        return .forwardStreamFailureAndFailTask(executor, error, nil)
+                    }
+
+                case .paused(let promises):
+                    switch result {
+                    case .success:
+                        self._state = .executing(executor, .finished, responseState)
+                        return .forwardStreamFinished(executor, promises)
+                    case .failure(let error):
+                        self._state = .finished(error: error)
+                        return .forwardStreamFailureAndFailTask(executor, error, promises)
+                    }
```

Comment:
Several of these compute on unwrapped `requestState`. Should they be `mutating func`s on the request state object itself?